Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp

```diff
@@ -45,10 +45,54 @@
 #import <pal/spi/ios/MobileGestaltSPI.h>
 #endif
 
+#if PLATFORM(VISION) && defined(__has_include) && __has_include(<libkern/OSThermalNotification.h>)
+#import <libkern/OSThermalNotification.h>
+#import <notify.h>
+#endif
+
 namespace WebKit {
 
+static Seconds adjustedTimeoutForThermalState(Seconds timeout)
+{
+#if PLATFORM(VISION) && defined(__has_include) && __has_include(<libkern/OSThermalNotification.h>)
+    static int thermalState = kOSThermalPressureLevelNominal;
+    static dispatch_once_t once = 0L;
+
+    dispatch_once(&once, ^{
+        int localToken;
+        uint32_t result = notify_register_dispatch(kOSThermalNotificationPressureLevelName, &localToken, dispatch_get_main_queue(), ^(int token) {
+            uint64_t localThermalState;
+            if (notify_get_state(token, &localThermalState) == NOTIFY_STATUS_OK)
+                thermalState = localThermalState;
+        });
+
+        uint64_t localThermalState;
+        if (result == NOTIFY_STATUS_OK && notify_get_state(localToken, &localThermalState) == NOTIFY_STATUS_OK)
+            thermalState = localThermalState;
+    });
+
+    switch (thermalState) {
+    case kOSThermalPressureLevelNominal:
+    case kOSThermalPressureLevelLight:
+    case kOSThermalPressureLevelModerate:
+        return timeout;
+    case kOSThermalPressureLevelHeavy:
+        return timeout * 5;
+    case kOSThermalPressureLevelTrapping:
+        return timeout * 10;
+    case kOSThermalPressureLevelSleeping:
+        return timeout;
+    }
+
+    ASSERT_NOT_REACHED("unexpected thermal level");
+    return timeout;
+#else
+    return timeout;
+#endif
+}
+
 AuxiliaryProcessProxy::AuxiliaryProcessProxy(bool alwaysRunsAtBackgroundPriority, Seconds responsivenessTimeout)
-    : m_responsivenessTimer(*this, responsivenessTimeout)
+    : m_responsivenessTimer(*this, adjustedTimeoutForThermalState(responsivenessTimeout))
```

Comment:
I'm not familiar with the mechanism via which the responsiveness timer will stop pending page loads -- what happens?

Reply:
Yes, I made a note in the commit message but I didn't want to make this a .mm file as it and its callers appear to be used in non-Cocoa contexts. I could create a new .mm file with a C-header and call it from here, that would also address the issue.

This code is mostly duplicated straight from NSProcessInfo however with some small customization.

Reply:
Well, I wonder if you should just use ThermalMitigationNotifier, which abstracts over the concept with a C++ API??

Reply:
@hortont424 I updated the PR to use `ThermalMitigationNotifier` instead so I don't have to import `libkern/OSThermalNotifcation.h` - is this acceptable now?